domaci_maze.py

```diff
@@ -90,495 +77,523 @@ def has_value(self) -> bool:
+            x += prob * (next_node.get_reward() + gamma * values[next_node.get_position()])
+        possible_values.append(x)
+    return max(possible_values) if max(possible_values) != 0 else -100  # wall node
 
-values = init_values(env)
-draw_values(env, values)
 
-nrows, ncols = 2, 2
-fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(15, 15))
-axes = axes.flatten()
-values = init_values(env)
-for k, ax in enumerate(axes):
-    draw_values(env, values, ax=ax)
-    ax.set_title(f"it={k}")
-    values = async_update_all_values(env, values, 1.0)
+# update q value for one node, action
+def update_q_value(env: MazeEnvironment, state, values, gamma):
+    current_node = env.get_current_pos_node(state[0])
+    possible_values = []
+    for next_node, prob in env.get_action_probabilities(current_node, state[1]):
+        possible_values_one_node = []
+        for action in ACTIONS:
+            possible_values_one_node.append(values[next_node.get_position(), action])
+        possible_values.append(prob * (next_node.get_reward() + gamma * max(possible_values_one_node)))
+    return sum(possible_values) if possible_values else -100  # wall node
 
 
-def value_iteration(env, gamma, eps, v0=None, maxiter=100):
-    v = v0 if v0 is not None else init_values(env)
-    for k in range(maxiter):
-        nv = async_update_all_values(env, values, gamma)
-        err = max([abs(nv[s] - v[s]) for s in v])
+# updates values for one iteration
+def async_update_all_values(env: MazeEnvironment, values, gamma, q_function):
+    for s in values:
+        if q_function:
+            if not env.is_terminal_pos(s[0]):  # q values are (position,action)
+                values[s] = update_q_value(env, s, values, gamma)
+        else:
+            if not env.is_terminal_pos(s):
+                values[s] = update_v_value(env, s, values, gamma)
+    return copy(values)
+
+
+def value_iteration(env, gamma, eps, maxit=100, q_function=False):
+    values = init_q_values(env) if q_function else init_v_values(env)
+    for iteration in range(maxit):
+        # print(f"\nIteration: {iteration + 1}")
+        # print("Old values: ", values)
+        values_copy = copy(values)
+        new_values = async_update_all_values(env, values, gamma, q_function)
+        # print("New values: ", new_values)
+        err = max([abs(new_values[s] - values_copy[s]) for s in values])
         if err < eps:
-            return nv, k
-        v = nv
-    return v, k
-
+            # print("Final error is: ", err)
+            return new_values, iteration + 1
+        # print("Error is: ", err)
+        values = new_values
+    return values, iteration + 1
+
+
+# should return the smallest number action if there are more actions that have the same value
+# if values of both actions 1 and 2 are -1.0 , it returns 1
+def best_action_min_arg(actions_probs):
+    max_probability = max(prob for _, prob in actions_probs)

```

Comment:
max(action_probs)

Reply:
A i to zapravo nisu probs nego values jel?

Reply:
Ova funckija je u sustini ni ne mora da se pozove, vidite ovaj zakomentarisani return u greedy_action, to je original. Taj zakomentarisani deo vraca akciju sa najvecom vrednosti dobiti. E sad ova funckija ima cilj da ako imaju vise akcija sa istom maksimalnom dobiti, da ona vrati manju ( npr akcije 3 4 i 5 imaju procenjenu dobit -2, ona ce uvek vratiti 3). Napravljena je samo i iskljucivo da bi se politike na kraju vise slagale ( jer bez nje moze jednom da izbaci 3 jednom 4 jednom 5 - cime se politike ne slazu, ali vrednost im je ista ).